### PR TITLE
[FIX] Fix of inconsistent indices #1726 #1731 #1732 #1734

### LINKED ISSUE #1724: Incorrect index of `MapReduce` operation at PandasQueryCompiler
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Modin version** (`modin.__version__`): 0.7.3+198.ge4e3fecb
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
if __name__ == "__main__":
    import modin.pandas as pd
    import numpy as np

    data = {"A": np.arange(256)}

    qc = pd.DataFrame(data)._query_compiler

    qc = qc.any()

    print("result:\n", qc.to_pandas())
    print("\nindex:\n", qc.index)
```

### Describe the problem

The output of the code above will be:
```
result:
       0
0  True

index:
 RangeIndex(start=0, stop=256, step=1)
```
As we see, index and data is inconsistent, API level is okay with this, because in that case it will call `to_pandas().squeeze()` to get a scalar. But it would useful to always know the correct index when you working at query compiler level.

### LINKED ISSUE #1726: to_pandas in PandasQueryCompiler loses meta information about indices and columns
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Modin version** (`modin.__version__`): 0.7.3+198.ge4e3fecb
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
if __name__ == "__main__":
    import modin.pandas as pd
    import numpy as np

    data = {"A": np.arange(256)}

    index = pd.MultiIndex.from_tuples((i, i*2) for i in np.arange(257)).drop(0)

    md_df = pd.DataFrame(data, index=index)

    print("md_df.index:", len(md_df.index.levels[0]))  # md_df.index: 257 
    print("pd_df.index:", len(md_df._to_pandas().index.levels[0]))  # pd_df.index: 256

```

### Describe the problem

`to_pandas` loses information about index levels which is useful for operations that operates with indices levels internally (like `unstack`)

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1726, #1731, #1732 and #1734 
- [x] tests added and passing

Initially, this PR was fixing that `to_pandas` in pandas backend does not consider information about indices/columns of modin frame, only from partitions. But after that fix many tests started to fail because `df_equals`, that uses in many tests, use `to_pandas` to convert modin frame to pandas and then compare it with an expected result. It revealed that many operations returns modin frame with incorrect indices, so this PR is also fixing all of them.

### CODE REVIEW

**Comment by dchigarev** [comment]
All check seems to be pass, except few that canceled because of timeout

File: modin/engines/ray/generic/io.py [hunk 1]
```diff
@@ -41,6 +41,6 @@ def func(df):
             df.to_sql(**kwargs)
             return pandas.DataFrame()
 
-        result = qc._modin_frame._fold_reduce(1, func)
+        result = qc._modin_frame._apply_full_axis(1, func)
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/engines/ray/generic/io.py:44]: Look at the `func` above that we pass here to `_fold_reduce`, it returns empty DataFrame, so the shape partitions changing (including indices and columns). `_fold_reduce` can't handle changing shape on both axis, when `_apply_full_axis` can.

File: modin/backends/pandas/query_compiler.py [hunk 2]
```diff
@@ -480,7 +480,9 @@ def is_monotonic_reduce(df):
             edge_case = monotonic_fn(pandas.Series(edges_list))
             return [common_case and edge_case]
 
-        return MapReduceFunction.register(is_monotonic_map, is_monotonic_reduce)(self)
+        return MapReduceFunction.register(
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:483]: good note, but I suppose that I should fix this in a separated PR

File: modin/engines/base/frame/data.py [hunk 3]
```diff
@@ -1353,7 +1357,8 @@ def to_pandas(self):
                 df = pandas.DataFrame(columns=self.columns)
             else:
                 df = pandas.DataFrame(columns=self.columns, index=self.index)
-        df.index.name = self.index.name
+        df.index = self.index
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: yeah, good point, fixed it in the last commit

File: modin/data_management/functions/binary_function.py [hunk 4]
```diff
@@ -39,6 +39,7 @@ def caller(query_compiler, other, *args, **kwargs):
                             axis,
                             lambda l, r: func(l, r.squeeze(), *args, **kwargs),
                             other._modin_frame,
+                            preserve_labels=False,
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/data_management/functions/binary_function.py]: fixed in the last commit

File: modin/data_management/functions/groupby_function.py [hunk 5]
```diff
@@ -115,12 +115,14 @@ def compute_reduce(df):
                 except ValueError:
                     return compute_reduce(df.copy())
 
-            if axis == 0:
-                new_columns = qc.columns
-                new_index = None
-            else:
-                new_index = query_compiler.index
-                new_columns = None
+            new_index = None
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/data_management/functions/groupby_function.py]: done

File: modin/backends/pandas/query_compiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/backends/pandas/query_compiler.py:483]: I don't think it will take a lot of time. If it's convenient to do it in another PR, I don't mind.

File: modin/engines/ray/generic/io.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/engines/ray/generic/io.py:44]: Okay, I see. Since we know that the result will be consist of empty partitions, may be, make sense pass to `_apply_full_axis` some values `new_index` and `new_columns` which are applied for empty dataframe?

[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/engines/ray/generic/io.py]: done

File: modin/backends/pandas/query_compiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:483]: fixed it in this PR

[the remark this answers is not in the document]
Comment (YarShev) [reply modin/backends/pandas/query_compiler.py:483]: Thanks!

**Comment by YarShev** [comment]
LGTM! Let's wait for @devin-petersohn , and @gshimansky .

File: modin/engines/ray/generic/io.py [hunk 6]
```diff
@@ -41,6 +41,6 @@ def func(df):
             df.to_sql(**kwargs)
             return pandas.DataFrame()
 
-        result = qc._modin_frame._fold_reduce(1, func)
+        result = qc._modin_frame._apply_full_axis(1, func, new_index=[], new_columns=[])
```
Comment (devin-petersohn) [inline modin/engines/ray/generic/io.py:44]: Why is `new_index=[]` and not `None`? Same question for `new_columns`

File: modin/pandas/groupby.py [hunk 7]
```diff
@@ -295,7 +295,7 @@ def __getitem__(self, key):
                 and all(c in self._columns for c in self._by.columns)
                 and self._drop
             ):
-                key = [key] + list(self._by.columns)
+                key = list(self._by.columns) + [key]
```
Comment (devin-petersohn) [inline modin/pandas/groupby.py:298]: Why is this change needed? It should not impact anything from what I see.

File: modin/engines/base/frame/data.py [hunk 8]
```diff
@@ -1353,7 +1357,9 @@ def to_pandas(self):
                 df = pandas.DataFrame(columns=self.columns)
             else:
                 df = pandas.DataFrame(columns=self.columns, index=self.index)
-        df.index.name = self.index.name
+        else:
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:1360]: We should fix the columns/index internally for these cases instead of making this change.

We should also add an `assert` statement here.

File: modin/data_management/functions/groupby_function.py [hunk 9]
```diff
@@ -115,19 +115,9 @@ def compute_reduce(df):
...
-                new_columns = None
+            # TODO: try to precompute `new_index` and `new_columns`
             new_modin_frame = qc._modin_frame.groupby_reduce(
-                axis,
-                by._modin_frame,
-                _map,
-                _reduce,
-                new_columns=new_columns,
-                new_index=new_index,
+                axis, by._modin_frame, _map, _reduce
```
Comment (devin-petersohn) [inline modin/data_management/functions/groupby_function.py:120]: Can we set `new_columns` and `new_index` based on some condition?

File: modin/engines/ray/generic/io.py (hunk 6 above)
Comment (dchigarev) [reply modin/engines/ray/generic/io.py:44]: We know that the result of applying `func` will be an empty DataFrame, so instead of extra twice calling `get_indices` in `_apply_full_axis` we specify new indices explicitly here

File: modin/pandas/groupby.py (hunk 7 above)
Comment (dchigarev) [reply modin/pandas/groupby.py:298]: this is the order in which pandas places columns at `groupby.__getitem__`, look at this reproducer:
```python
if __name__ == "__main__":
    import modin.pandas as pd
    import pandas
    import numpy as np

    data = {
        "col1": [0, 1, 2, 3],
        "col2": [4, 5, np.NaN, 7],
        "col3": [np.NaN, np.NaN, 12, 10],
        "col4": [17, 13, 16, 15],
        "col5": [-4, -5, -6, -7],
    }

    pd_grp = pandas.DataFrame(data).groupby("col4", as_index=False)
    md_grp = pd.DataFrame(data).groupby("col4", as_index=False)

    print("pd_columns:", pd_grp["col1"].count().columns)   # pd_columns: Index(['col4', 'col1'], dtype='object')
    print("md_columns:", md_grp["col1"].count().columns)  # md_columns: Index(['col1', 'col4'], dtype='object') 
```

File: modin/data_management/functions/groupby_function.py (hunk 9 above)
Comment (dchigarev) [reply modin/data_management/functions/groupby_function.py:120]: Since we don't know what passed function would do with our DataFrame I'm afraid not. Let's remember our approach from #1645 where applied function transforms and changes shape of DataFrames in partitions. If we allow to apply functions like this, it's impossible to predict new indices/columns before computing.

File: modin/pandas/groupby.py (hunk 7 above)
Comment (devin-petersohn) [reply modin/pandas/groupby.py:298]: Yes, I see.

File: modin/engines/base/frame/data.py (hunk 8 above)
Comment (dchigarev) [reply modin/engines/base/frame/data.py:1360]: Look at the original reproducer of #1726, even if partitions contains correct indices, `concat` that used in `to_pandas` will lose information about levels, so I believe we should add `assert` statement for equality internal and external indices values and after that do that assignment
```python
df.index = self.index
df.columns = self.columns
```

Comment (dchigarev) [reply modin/engines/base/frame/data.py:1360]: added assert check in the last commit

**Comment by dchigarev** [comment]
@devin-petersohn I guess I addressed all your comments, please take a look. I've also added propagation indices names like `["__reduced__"]` for Series into partitions, to it successfully passed assert check in `to_pandas`

File: modin/engines/base/frame/data.py [hunk 10]
```diff
@@ -74,6 +78,8 @@ def __init__(
         self._column_widths_cache = column_widths
         self._dtypes = dtypes
         self._filter_empties()
+        if validate_axes is not None and validate_axes is not False:
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:81]: Can `validate_axes` be `None`? We can just give a type hint in the method signature for `validate_axes: bool = False`.

File: modin/engines/base/frame/data.py [hunk 11]
```diff
@@ -234,6 +246,64 @@ def _filter_empties(self):
         self._column_widths_cache = [w for w in self._column_widths if w > 0]
         self._row_lengths_cache = [r for r in self._row_lengths if r > 0]
 
+    def _validate_axis_equality(self, axis, update):
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:249]: Can we use type hints and add some method level documentation as well?

File: modin/engines/base/frame/data.py [hunk 12]
```diff
@@ -234,6 +246,64 @@ def _filter_empties(self):
...
+                not is_equals, "Internal and external indices do not match.",
+            )
+        elif not is_equals:
+            self._set_axis(axis, self.axes[axis])
+
+    def _validate_internal_indices(self, mode=None, **kwargs):
+        """
+        Validates and optionally updates internal and external indices
+        of modin_frame in specified mode. There is 3 modes supported:
+            1. "reduced only" - validates and updates indices on that axes
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:265]: Instead of `"reduced only"` can we use `"reduced"` or something similar? (just a single word is preferred).

File: modin/engines/base/frame/data.py (hunk 11 above)
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: added docstring and type hints

File: modin/engines/base/frame/data.py (hunk 10 above)
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: Yeah, you're right, `validate_axes` is not supposed to be `None`, added type hint here

File: modin/engines/base/frame/data.py (hunk 12 above)
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: changed it to `"reduced"`

**Comment by dchigarev** [comment]
@devin-petersohn comments addressed and CI is green (for the exception of `black` check that seems to be frozen)

File: modin/engines/base/frame/data.py [hunk 13]
```diff
@@ -204,6 +211,21 @@ def _set_columns(self, new_columns):
                 self._dtypes.index = new_columns
         self._apply_index_objs(axis=1)
 
+    def _set_axis(self, axis, new_axis):
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:214]: I prefer not to add this, you can use `self.axes[axis] = new_axis`

File: modin/engines/base/frame/data.py [hunk 14]
```diff
@@ -234,6 +256,74 @@ def _filter_empties(self):
...
+            axis : int,
+                Axis to validate indices along
+            update : bool,
+                Update or raise an exception if indices don't match
+        """
+        internal_axis = self._frame_mgr_cls.get_indices(
+            axis, self._partitions, lambda df: df.axes[axis]
+        )
+        is_equals = self.axes[axis].equals(internal_axis)
+        if not update:
```
Comment (devin-petersohn) [inline modin/engines/base/frame/data.py:274]: I did not see this getting used in this PR, how are you expecting this to be used?

Comment (devin-petersohn) [reply modin/engines/base/frame/data.py:274]: It may be better to just leave the `update` parameter out because I don't think we ever want to throw the internal error. What do you think?

File: modin/engines/base/frame/data.py (hunk 13 above)
Comment (dchigarev) [reply modin/engines/base/frame/data.py:214]: `self.axes` will return us `[self.index, self.columns]`, roughly say just an array of values, so assignment like `self.axes[axis] = new_axis` will not do any effect, look at that reproduce:
```python
import pandas

df = pandas.DataFrame([1,2,3])
print(df.axes[0]) # RangeIndex(start=0, stop=3, step=1)
df.axes[0] = pandas.Index(['a', 'b', 'c'])
print(df.axes[0]) # RangeIndex(start=0, stop=3, step=1)
```

File: modin/engines/base/frame/data.py (hunk 14 above)
Comment (dchigarev) [reply modin/engines/base/frame/data.py]: well, if you think so

removed that parameter in the last commit

**Comment by dchigarev** [comment]
@devin-petersohn I've found out that `df_equals` [sorts columns labels](https://github.com/modin-project/modin/blob/master/modin/pandas/test/utils.py#L458-L466) before asserting frame equality, so after I've added assertion for columns equality (without that sorting) many tests appear to be failed. Are we really want to sort columns to check frames equality or this is a legacy stuff?

File: modin/engines/base/frame/data.py (hunk 13 above)
Comment (devin-petersohn) [reply modin/engines/base/frame/data.py:214]: Not enough coffee yet 😄

**Comment by devin-petersohn** [comment]
>Are we really want to sort columns to check frames equality or this is a legacy stuff?

This is legacy, we should not be sorting. This is a different issue, let's open a new issue and PR for that and add the check in `df_equals` there. 

It looks like CI is failing to download the packages.

**Comment by dchigarev** [comment]
> > Are we really want to sort columns to check frames equality or this is a legacy stuff?
> 
> This is legacy, we should not be sorting. This is a different issue, let's open a new issue and PR for that and add the check in `df_equals` there.
> 
> It looks like CI is failing to download the packages.

removed asserting columns from that branch and created a separate issue about sorting labels (#1784)

**Comment by dchigarev** [comment]
@devin-petersohn CI is green, I guess PR is ready to merge
